### PR TITLE
attempt to fix caching

### PR DESCRIPTION
**Pull request by pkhry**
### Description 

attempts to fix caching for `resolc`. 

This PR will also:
- revert the logging code for the `Reporter`, because we now print out that we depend on `solc` in version.build.
   e.g `[⠘] Compiling 24 files with Resolc 0.1.0-dev.15+Solc.0.8.29`

see for explanations https://github.com/paritytech/foundry-polkadot/issues/125. 

### Issues that still persist 
forge build will recompile only once with the given new version and stop afterwards. 
e.g. 
```bash
$ forge build --resolc-compile --use_resolc=0.1.0-dev.13
[⠊] Compiling...
<compiled successfully>
$ forge build --resolc-compile --use_resolc=0.1.0-dev.16
[⠊] Compiling...
<compiled successfully>
$ forge build --resolc-compile --use_resolc=0.1.0-dev.13
[⠊] Compiling...
No files changed, compilation skipped
```

The same issue is also encountered when using plain `solc`.

### CODE REVIEW

File: crates/compilers/src/cache.rs [hunk 1]
```diff
@@ -187,6 +187,8 @@ impl<S: CompilerSettings> CompilerCache<S> {
                     .builds
                     .contains(build_id.file_name().to_string_lossy().trim_end_matches(".json"))
                 {
+                    self.builds
+                        .remove(build_id.file_name().to_string_lossy().trim_end_matches(".json"));
                     let _ = std::fs::remove_file(build_id.path());
```
Comment (filip-parity) [inline crates/compilers/src/cache.rs:192]: **(nitpick)** Add error handling or logging for cases where `remove_file` or `builds.remove` fails, and validate `build_id` filename assumptions in tests.

File: crates/compilers/src/compilers/resolc/compiler.rs [hunk 2]
```diff
@@ -35,7 +35,9 @@ impl Compiler for Resolc {
     type Language = SolcLanguage;
 
     fn compiler_version(&self, _input: &Self::Input) -> Version {
-        self.resolc_version.clone()
+        let mut v = self.resolc_version.clone();
+        v.build = semver::BuildMetadata::new(&format!("Solc.{}", _input.version())).unwrap();
```
Comment (filip-parity) [inline crates/compilers/src/compilers/resolc/compiler.rs:39]: **(nitpick)** Add error handling for `BuildMetadata::new` and test edge cases (e.g., malformed `solc` versions).

File: crates/compilers/src/report/mod.rs
Comment (filip-parity) [inline crates/compilers/src/report/mod.rs:1]: **(question)** This removal assumes all use cases are covered by single-version reporting, correct? If future compilers require multi-version reporting, this change may need to be reverted or extended. **Are we expecting this?**

File: crates/compilers/src/compilers/resolc/compiler.rs [hunk 3]
```diff
@@ -75,7 +77,7 @@ impl Compiler for Resolc {
 
 impl SimpleCompilerName for Resolc {
     fn compiler_name_default() -> std::borrow::Cow<'static, str> {
-        "Resolc and Solc".into()
+        "Resolc".into()
```
Comment (filip-parity) [inline crates/compilers/src/compilers/resolc/compiler.rs:80]: **(nitpick)** Add a comment explaining why `"Resolc"` is sufficient.

**Review by filip-parity — COMMENTED** [review]
**LGTM**, a couple of nitpicks (_feel free to skip them if you prefer_) and a question.

File: crates/compilers/src/cache.rs (hunk 1 above)
Comment (pkhry) [reply crates/compilers/src/cache.rs:192]: builds remove was added by mistake, removed it. remove_file is just written as in above code without any error handling

File: crates/compilers/src/compilers/resolc/compiler.rs (hunk 2 above)
Comment (pkhry) [reply crates/compilers/src/compilers/resolc/compiler.rs]: can not happen according to the parsing code. updated code will use `version.{major/minor/patch} ` directly to avoid potential errors

File: crates/compilers/src/report/mod.rs
Comment (pkhry) [reply crates/compilers/src/report/mod.rs:1]: I don't think we will be printing more than one direct dependency for a given compiler. in case of `resolc` it is `solc`

**Review by filip-parity — APPROVED** [review]

File: crates/compilers/src/compilers/resolc/compiler.rs [hunk 4]
```diff
@@ -35,7 +35,17 @@ impl Compiler for Resolc {
     type Language = SolcLanguage;
 
     fn compiler_version(&self, _input: &Self::Input) -> Version {
-        self.resolc_version.clone()
+        let mut v = self.resolc_version.clone();
```
Comment (smiasojed) [inline crates/compilers/src/compilers/resolc/compiler.rs:38]: If I understand this correctly, now we combine resolc with solc version, so why in hasher we need to hash input and compiler_version?

Comment (pkhry) [reply crates/compilers/src/compilers/resolc/compiler.rs:38]: hmm ,good call. I dunno what to do  with `short_version` in the build_info hash tho, for removal as well as it might not work properly?

File: crates/compilers/src/compilers/resolc/compiler.rs [hunk 5]
```diff
@@ -35,7 +35,17 @@ impl Compiler for Resolc {
...
+        let input_version = _input.version();
+
+        // Note it shouldn't fail as parsing code assumes that there can be an optional string
+        // that precludes the version number
+        v.build = semver::BuildMetadata::new(&format!(
+            "Solc.{}.{}.{}",
+            input_version.major, input_version.minor, input_version.patch
+        ))
+        .expect("Can't fail");
+        v
```
Comment (smiasojed) [inline crates/compilers/src/compilers/resolc/compiler.rs:48]: I am not sure if this concat is a good idea

Comment (pkhry) [reply crates/compilers/src/compilers/resolc/compiler.rs:48]: care to elaborate?

Comment (smiasojed) [reply crates/compilers/src/compilers/resolc/compiler.rs:48]: BuildMetadata if they are present will not match with the real compiler version

Comment (pkhry) [reply crates/compilers/src/compilers/resolc/compiler.rs:48]: > BuildMetadata if they are present will not match with the real compiler version, because we put here solidity version

are there plans to add build_metadata to resolc version output ?

Comment (smiasojed) [reply crates/compilers/src/compilers/resolc/compiler.rs:48]: If I saw such a version and tried to find the corresponding compiler from a release, I’d be in trouble — because that release wouldn't exist and could change with every contract compilation, causing confusion.

I'm also not sure if forge resolve would work in this case, or whether any of our tests in foundry-polkadot would still pass.

Comment (pkhry) [reply crates/compilers/src/compilers/resolc/compiler.rs:48]: tests did in fact pass after a small regex change, why would `resolve` not work?

Comment (smiasojed) [reply crates/compilers/src/compilers/resolc/compiler.rs:48]: I'm not a big fan of this approach—it feels a bit hacky. Is there another option we could consider? That said, I don't have a strong opinion.

**Comment by smiasojed** [comment]
Could we please separate the PR for caching from the one for compiler version reporting? Caching is very important for us, whereas compiler reporting is not a priority.

**Comment by pkhry** [comment]
> Could we please separate the PR for caching from the one for compiler version reporting? Caching is very important for us, whereas compiler reporting is not a priority.

those two are intertwined in the PR

File: crates/compilers/src/compilers/resolc/compiler.rs (hunk 5 above)
Comment (pkhry) [reply crates/compilers/src/compilers/resolc/compiler.rs:48]: how about filing an issue and introducing proper `CompilerInfo` trait instead of just using `Input::Version`?

Comment (smiasojed) [reply crates/compilers/src/compilers/resolc/compiler.rs:48]: full revive version is : `0.1.0-dev.16+commit.0421869.llvm-18.1.8`
solc: `0.8.29+commit.e7d22988.Darwin.appleclang`
We need to put them both to Json compiler output
I will take a look on it

Comment (pkhry) [reply crates/compilers/src/compilers/resolc/compiler.rs:48]: i can implement the `CmpilerInfo` struct tomorrow if needed

File: crates/compilers/src/compile/project.rs [hunk 6]
```diff
@@ -644,6 +648,29 @@ fn compile_parallel<'a, C: Compiler>(
     })
 }
 
+fn compound_version(mut compiler_version: Version, input_version: &Version) -> Version {
```
[the remark this answers is not in the document]
Comment (pkhry) [reply crates/compilers/src/compile/project.rs:651]: too much refactoring for now, the whole cache structure assumes you use a semver::Version for keys and we have usages outside of cache

**Comment by smiasojed** [comment]
I looks that it works, thanks

**Review by smiasojed — APPROVED** [review]
